### ISSUE TITLE
SqlLogger.logException() strange behavior

### ISSUE THREAD

**Issue by vlaate** [jdbi/jdbi#1726]
I implemented a trivial SqlLogger that just prints messages to System.out (before, after, and on exception)

My intention was to better understand the behavior around SQLLogger, and particularly see if after the call to **logException** (featuring a SQLException argument) the exception was re-thrown by jdbi (which I'd find desirable).

I was surprised:

When I test it with correct queries, I see that both **logBeforeExecution** and **logAfterExecution** are invoked and my messages get printed.

However, when I break the sql in any way, so as to produce an Exception, I don't get any of my messages printed, not for **logException** but also not for **logBeforeExecution** either, which is what I found most strange. 

I'm using SLF4j-simple, and I still get a print of a stacktrace on System.err (which is desirable), and the stacktrace says "UnableToCreateStatementException" at SqlStatement.internalExecute, and includes a cause exception from HQSLDB about whatever I broke on the SQL.

From the stacktrace and checking the jdbi code, I think the error is in:
[https://github.com/jdbi/jdbi/blob/master/core/src/main/java/org/jdbi/v3/core/statement/SqlStatement.java#L1656](https://github.com/jdbi/jdbi/blob/master/core/src/main/java/org/jdbi/v3/core/statement/SqlStatement.java#L1656)

It throws an **UnableToCreateStatementException** at line 1656 before the SqlLogger gets invoked in line 1668:

```
        try {
            stmt = createStatement(ctx, parsedSql);
            // The statement builder might (or might not) clean up the statement when called. E.g. the
            // caching statement builder relies on the statement *not* being closed.
            addCleanable(() -> handle.getStatementBuilder().close(handle.getConnection(), this.sql, stmt));
            getConfig(SqlStatements.class).customize(stmt);
        } catch (SQLException e) {
            throw new UnableToCreateStatementException(e, ctx);
        }
```

I think that on SqlStatement line 1656, before throwing the **UnableToCreateStatementException**, a call should be made to the SqlLogger.logException. ... or perhaps wrap the statement with SqlLoggerUtil.wrap like it's don't in line 1668?

I'm sorry I can't help by pushing a tested fix for this.

**Comment by leaumar** [comment]
High-level-logically, I personally wouldn't expect something that logs _sql statements_ to be involved when the error is something as early as that _an sql statement could not be built_, i.e. the subject to log does not exist. SqlLogger is meant to be a record of operations effectively performed by the database.

**Comment by leaumar** [comment]
What would you expect it to be called with in this case? The query that couldn't be "compiled" into correct sql?

**Comment by vlaate** [comment]
> What would you expect it to be called with in this case? The query that couldn't be "compiled" into correct sql?

If instead of breaking the SQL, I remove read privileges on the table for this user, the same thing happens.
If instead of breaking the SQL I drop the table on the database server, the same thing happens.

From your first comment I now see the concept of **logBeforeExecution** might actually mean "before _jdbc statement_ execution", and that would make sense. It's just that with the interface being called "SqlLogger", one might interpret "before execution" to mean "before execution of the SQL" such as before using it to create a jdbc statement.

If one wants to instrument monitoring of a JDBI app, not getting **logBeforeExecution** nor **logException** called to detect when access privileges to the table involved have been changed in production, might pose a problem.

**Comment by stevenschlansker** [comment]
Yes, sorry this isn't clear - but `beforeExecution` does exactly mean before *statement* execution.  If the statement cannot be created, then it is never ready to execute, and so the "before" hook never fires.

SqlLogger is meant to provide some useful loggable information, but it is not comprehensive and does not reflect all problems that might happen.  Many configuration or preparation problems, such as an argument failing to bind, or unparseable SQL, will not get far enough to trigger this callback.

Regarding instrumenting, the fact that Jdbi throws an exception should be enough to determine the problem.  Either you catch and handle the exception, or log it via your slf4 logger as a severe problem.  Is there some reason the thrown exception isn't an appropriate way for your application to learn of the problem?

I don't think we can change the current behavior dramatically without risking breaking other users.  I do think we can make the documentation more clear about exactly when the callbacks will or won't get called.
